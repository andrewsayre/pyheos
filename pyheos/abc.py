"""Define abstract base classes for HEOS."""

from abc import ABC
from typing import Any


class RemoveHeosFieldABC(ABC):
    """Define an abstract base class that removes the 'heos' from dataclass's fields list to prevent serialization."""

    def __post_init__(self, *args: Any, **kwargs: Any) -> None:
        """Post initialize the player."""
        # Prevent the heos instance from being serialized
        fields = self.__dataclass_fields__.copy()  # type: ignore[has-type] # pylint: disable=access-member-before-definition
        del fields["heos"]
        self.__dataclass_fields__ = fields
