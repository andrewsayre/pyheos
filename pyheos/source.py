"""Define the heos source module."""
from typing import Optional, Sequence  # pylint: disable=unused-import


class InputSource:
    """Define an input source."""

    def __init__(self, player_id: int, name: str, input_name: str):
        """Init the source."""
        self._player_id = player_id  # type: int
        self._name = name  # type: str
        self._input_name = input_name  # type: str

    def __str__(self):
        """Get a user-readable representation of the source."""
        return "<{} ({})>".format(self._name, self._input_name)

    def __repr__(self):
        """Get a debug representation of the source."""
        return "<{} ({}) on {}>".format(self._name, self._input_name,
                                        self._player_id)

    @property
    def name(self) -> str:
        """Get the friendly display name."""
        return self._name

    @property
    def input_name(self) -> str:
        """Get the input source name."""
        return self._input_name

    @property
    def player_id(self) -> int:
        """Get the player id."""
        return self._player_id


class HeosSource:
    """Define an individual heos source."""

    def __init__(self, commands, data: Optional[dict] = None):
        """Init the source class."""
        self._commands = commands
        self._name = None  # type: str
        self._image_url = None  # type: str
        self._type = None  # type: str
        self._source_id = None  # type: int
        self._available = None  # type: bool
        self._service_username = None  # type: str
        self._container = None  # type: bool
        self._media_id = None  # type: str
        self._playable = None  # type: bool
        if data:
            self._from_data(data)

    def _from_data(self, data: dict):
        self._name = data['name']
        self._image_url = data['image_url']
        self._type = data['type']

        source_id = data.get('sid')
        if source_id:
            self._source_id = int(source_id)

        self._available = data.get('available') == 'true'
        self._service_username = data.get('service_username')
        self._container = data.get('container') == 'yes'
        self._media_id = data.get('mid')
        self._playable = data.get('playable') == 'yes'

    def __str__(self):
        """Get a user-readable representation of the source."""
        return "<{} ({})>".format(self._name, self._type)

    def __repr__(self):
        """Get a debug representation of the source."""
        return "<{} ({}) {}>".format(self._name, self._type, self._source_id)

    async def browse(self) -> 'Sequence[HeosSource]':
        """Browse the contents of the current source."""
        items = await self._commands.browse(self._source_id)
        return [HeosSource(self._commands, item) for item in items]

    @property
    def name(self) -> str:
        """Get the name of the source."""
        return self._name

    @property
    def image_url(self) -> str:
        """Get the image url of the source."""
        return self._image_url

    @property
    def type(self) -> str:
        """Get the type of the source."""
        return self._type

    @property
    def source_id(self) -> int:
        """Get the id of the source."""
        return self._source_id

    @property
    def available(self) -> bool:
        """Return True if the source is available."""
        return self._available

    @property
    def service_username(self) -> str:
        """Get the service username."""
        return self._service_username

    @property
    def media_id(self) -> str:
        """Get the media id."""
        return self._media_id

    @property
    def container(self) -> bool:
        """Return True if the source is a container."""
        return self._container

    @property
    def playable(self) -> bool:
        """Return True if the source is playable."""
        return self._playable
