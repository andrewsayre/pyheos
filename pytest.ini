[pytest]
log_level=DEBUG
testpaths = tests
norecursedirs = .git
timeout = 60
asyncio_default_fixture_loop_scope = function