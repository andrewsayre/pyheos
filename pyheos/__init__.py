"""pyheos - a library for interacting with HEOS devices."""
from . import const
from .dispatch import Dispatcher
from .heos import Heos
from .player import HeosNowPlayingMedia, HeosPlayer
from .response import CommandError
from .source import HeosSource, InputSource

__all__ = [
    'const',
    'CommandError',
    'Dispatcher',
    'Heos',
    'HeosPlayer',
    'HeosNowPlayingMedia',
    'HeosSource',
    'InputSource'
]
