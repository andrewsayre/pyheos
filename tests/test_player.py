"""Tests for the player module."""
import pytest

from pyheos import const
from pyheos.player import HeosPlayer
from pyheos.source import InputSource


def test_str():
    """Test the __str__ function."""
    data = {
        "name": "Back Patio",
        "pid": 1,
        "model": "HEOS Drive",
        "version": "1.493.180",
        "ip": "192.168.0.1",
        "network": "wired",
        "lineout": 1
    }
    player = HeosPlayer(None, data)
    assert str(player) == '{Back Patio (HEOS Drive)}'
    assert repr(player) == '{Back Patio (HEOS Drive) with id 1 at 192.168.0.1}'


@pytest.mark.asyncio
async def test_set_state(mock_device, heos):
    """Test the play, pause, and stop commands."""
    await heos.get_players()
    player = heos.get_player(1)

    mock_device.register(const.COMMAND_SET_PLAY_STATE, None,
                         'player.set_play_state')

    assert await player.play()
    assert await player.pause()
    assert await player.stop()

    with pytest.raises(ValueError):
        await player.set_state("invalid")


@pytest.mark.asyncio
async def test_set_volume_and_mute(mock_device, heos):
    """Test the volume commands."""
    await heos.get_players()
    player = heos.get_player(1)

    mock_device.register(const.COMMAND_SET_VOLUME, None, 'player.set_volume')
    mock_device.register(const.COMMAND_SET_MUTE, None, 'player.set_mute')
    mock_device.register(const.COMMAND_VOLUME_UP, None, 'player.volume_up')
    mock_device.register(const.COMMAND_VOLUME_DOWN, None, 'player.volume_down')
    mock_device.register(const.COMMAND_TOGGLE_MUTE, None, 'player.toggle_mute')

    # Volume
    with pytest.raises(ValueError):
        await player.set_volume(-1)
    with pytest.raises(ValueError):
        await player.set_volume(101)
    assert await player.set_volume(100)

    # Mute
    assert await player.mute()
    assert await player.unmute()
    assert await player.toggle_mute()

    # Up
    assert await player.volume_up(6)
    with pytest.raises(ValueError):
        await player.volume_up(0)
    with pytest.raises(ValueError):
        await player.volume_up(11)

    # Down
    assert await player.volume_down(6)
    with pytest.raises(ValueError):
        await player.volume_down(0)
    with pytest.raises(ValueError):
        await player.volume_down(11)


@pytest.mark.asyncio
async def test_set_play_mode(mock_device, heos):
    """Test the volume commands."""
    await heos.get_players()
    player = heos.get_player(1)
    args = {
        'pid': '1',
        'repeat': const.REPEAT_ON_ALL,
        'shuffle': 'on'
    }
    mock_device.register(const.COMMAND_SET_PLAY_MODE, args,
                         'player.set_play_mode')

    assert await player.set_play_mode(const.REPEAT_ON_ALL, True)
    # Assert invalid mode
    with pytest.raises(ValueError):
        await player.set_play_mode("repeat", True)


@pytest.mark.asyncio
async def test_play_next_previous(mock_device, heos):
    """Test the volume commands."""
    await heos.get_players()
    player = heos.get_player(1)
    args = {'pid': '1'}
    mock_device.register(const.COMMAND_PLAY_NEXT, args, 'player.play_next')
    mock_device.register(const.COMMAND_PLAY_PREVIOUS, args,
                         'player.play_previous')

    assert await player.play_next()
    assert await player.play_previous()


@pytest.mark.asyncio
async def test_clear_queue(mock_device, heos):
    """Test the volume commands."""
    await heos.get_players()
    player = heos.get_player(1)
    args = {'pid': '1'}
    mock_device.register(const.COMMAND_CLEAR_QUEUE, args, 'player.clear_queue')

    assert await player.clear_queue()

    # Also test with a 'command under process' response
    mock_device.register(const.COMMAND_CLEAR_QUEUE, args,
                         ['player.clear_queue',
                          'player.clear_queue_processing'], replace=True)

    assert await player.clear_queue()


@pytest.mark.asyncio
async def test_play_input_source(mock_device, heos):
    """Test the play input source."""
    await heos.get_players()
    player = heos.get_player(1)
    input_source = InputSource(1, "AUX In 1", const.INPUT_AUX_IN_1)
    args = {'pid': '1', 'spid': str(input_source.player_id),
            'input': input_source.input_name}
    mock_device.register(const.COMMAND_BROWSE_PLAY_INPUT, args,
                         'browse.play_input')

    assert await player.play_input_source(input_source)

    # Test invalid input_name
    with pytest.raises(ValueError):
        await player.play_input("Invalid")


@pytest.mark.asyncio
async def test_play_favorite(mock_device, heos):
    """Test the play favorite."""
    await heos.get_players()
    player = heos.get_player(1)
    args = {'pid': '1', 'preset': '1'}
    mock_device.register(const.COMMAND_BROWSE_PLAY_PRESET, args,
                         'browse.play_preset')

    assert await player.play_favorite(1)

    # Test invalid starting index
    with pytest.raises(ValueError):
        await player.play_favorite(0)


@pytest.mark.asyncio
async def test_play_url(mock_device, heos):
    """Test the play favorite."""
    await heos.get_players()
    player = heos.get_player(1)
    url = "https://my.website.com/podcast.mp3"
    args = {'pid': '1', 'url': url}
    mock_device.register(const.COMMAND_BROWSE_PLAY_STREAM,
                         args, 'browse.play_stream')

    assert await player.play_url(url)
