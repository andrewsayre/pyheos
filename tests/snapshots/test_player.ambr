# serializer version: 1
# name: test_from_data_ip_address[192.168.0.1]
  dict({
    'available': True,
    'control': <VolumeControlType.UNKNOWN: 0>,
    'group_id': None,
    'ip_address': '192.168.0.1',
    'is_muted': False,
    'line_out': <LineOutLevelType.VARIABLE: 1>,
    'model': 'HEOS Drive',
    'name': 'Back Patio',
    'network': <NetworkType.WIRED: 'wired'>,
    'now_playing_media': dict({
      'album': None,
      'album_id': None,
      'artist': None,
      'current_position': None,
      'current_position_updated': None,
      'duration': None,
      'image_url': None,
      'media_id': None,
      'options': list([
      ]),
      'queue_id': None,
      'song': None,
      'source_id': None,
      'station': None,
      'supported_controls': list([
      ]),
      'type': None,
    }),
    'playback_error': None,
    'player_id': 1,
    'repeat': <RepeatType.OFF: 'off'>,
    'serial': '1234567890',
    'shuffle': False,
    'state': None,
    'supported_version': True,
    'version': '3.34.0',
    'volume': 0,
  })
# ---
# name: test_from_data_ip_address[None]
  dict({
    'available': True,
    'control': <VolumeControlType.UNKNOWN: 0>,
    'group_id': None,
    'ip_address': None,
    'is_muted': False,
    'line_out': <LineOutLevelType.VARIABLE: 1>,
    'model': 'HEOS Drive',
    'name': 'Back Patio',
    'network': <NetworkType.WIRED: 'wired'>,
    'now_playing_media': dict({
      'album': None,
      'album_id': None,
      'artist': None,
      'current_position': None,
      'current_position_updated': None,
      'duration': None,
      'image_url': None,
      'media_id': None,
      'options': list([
      ]),
      'queue_id': None,
      'song': None,
      'source_id': None,
      'station': None,
      'supported_controls': list([
      ]),
      'type': None,
    }),
    'playback_error': None,
    'player_id': 1,
    'repeat': <RepeatType.OFF: 'off'>,
    'serial': '1234567890',
    'shuffle': False,
    'state': None,
    'supported_version': True,
    'version': '3.34.0',
    'volume': 0,
  })
# ---
# name: test_from_data_ip_address[]
  dict({
    'available': True,
    'control': <VolumeControlType.UNKNOWN: 0>,
    'group_id': None,
    'ip_address': '',
    'is_muted': False,
    'line_out': <LineOutLevelType.VARIABLE: 1>,
    'model': 'HEOS Drive',
    'name': 'Back Patio',
    'network': <NetworkType.WIRED: 'wired'>,
    'now_playing_media': dict({
      'album': None,
      'album_id': None,
      'artist': None,
      'current_position': None,
      'current_position_updated': None,
      'duration': None,
      'image_url': None,
      'media_id': None,
      'options': list([
      ]),
      'queue_id': None,
      'song': None,
      'source_id': None,
      'station': None,
      'supported_controls': list([
      ]),
      'type': None,
    }),
    'playback_error': None,
    'player_id': 1,
    'repeat': <RepeatType.OFF: 'off'>,
    'serial': '1234567890',
    'shuffle': False,
    'state': None,
    'supported_version': True,
    'version': '3.34.0',
    'volume': 0,
  })
# ---
# name: test_from_data_network[None]
  dict({
    'available': True,
    'control': <VolumeControlType.UNKNOWN: 0>,
    'group_id': None,
    'ip_address': '192.168.0.1',
    'is_muted': False,
    'line_out': <LineOutLevelType.VARIABLE: 1>,
    'model': 'HEOS Drive',
    'name': 'Back Patio',
    'network': <NetworkType.UNKNOWN: 'unknown'>,
    'now_playing_media': dict({
      'album': None,
      'album_id': None,
      'artist': None,
      'current_position': None,
      'current_position_updated': None,
      'duration': None,
      'image_url': None,
      'media_id': None,
      'options': list([
      ]),
      'queue_id': None,
      'song': None,
      'source_id': None,
      'station': None,
      'supported_controls': list([
      ]),
      'type': None,
    }),
    'playback_error': None,
    'player_id': 1,
    'repeat': <RepeatType.OFF: 'off'>,
    'serial': '1234567890',
    'shuffle': False,
    'state': None,
    'supported_version': True,
    'version': '3.34.0',
    'volume': 0,
  })
# ---
# name: test_from_data_network[invalid]
  dict({
    'available': True,
    'control': <VolumeControlType.UNKNOWN: 0>,
    'group_id': None,
    'ip_address': '192.168.0.1',
    'is_muted': False,
    'line_out': <LineOutLevelType.VARIABLE: 1>,
    'model': 'HEOS Drive',
    'name': 'Back Patio',
    'network': <NetworkType.UNKNOWN: 'unknown'>,
    'now_playing_media': dict({
      'album': None,
      'album_id': None,
      'artist': None,
      'current_position': None,
      'current_position_updated': None,
      'duration': None,
      'image_url': None,
      'media_id': None,
      'options': list([
      ]),
      'queue_id': None,
      'song': None,
      'source_id': None,
      'station': None,
      'supported_controls': list([
      ]),
      'type': None,
    }),
    'playback_error': None,
    'player_id': 1,
    'repeat': <RepeatType.OFF: 'off'>,
    'serial': '1234567890',
    'shuffle': False,
    'state': None,
    'supported_version': True,
    'version': '3.34.0',
    'volume': 0,
  })
# ---
# name: test_from_data_network[wired]
  dict({
    'available': True,
    'control': <VolumeControlType.UNKNOWN: 0>,
    'group_id': None,
    'ip_address': '192.168.0.1',
    'is_muted': False,
    'line_out': <LineOutLevelType.VARIABLE: 1>,
    'model': 'HEOS Drive',
    'name': 'Back Patio',
    'network': <NetworkType.WIRED: 'wired'>,
    'now_playing_media': dict({
      'album': None,
      'album_id': None,
      'artist': None,
      'current_position': None,
      'current_position_updated': None,
      'duration': None,
      'image_url': None,
      'media_id': None,
      'options': list([
      ]),
      'queue_id': None,
      'song': None,
      'source_id': None,
      'station': None,
      'supported_controls': list([
      ]),
      'type': None,
    }),
    'playback_error': None,
    'player_id': 1,
    'repeat': <RepeatType.OFF: 'off'>,
    'serial': '1234567890',
    'shuffle': False,
    'state': None,
    'supported_version': True,
    'version': '3.34.0',
    'volume': 0,
  })
# ---
# name: test_from_data_version[1.493.180]
  dict({
    'available': True,
    'control': <VolumeControlType.UNKNOWN: 0>,
    'group_id': None,
    'ip_address': '192.168.0.1',
    'is_muted': False,
    'line_out': <LineOutLevelType.VARIABLE: 1>,
    'model': 'HEOS Drive',
    'name': 'Back Patio',
    'network': <NetworkType.WIRED: 'wired'>,
    'now_playing_media': dict({
      'album': None,
      'album_id': None,
      'artist': None,
      'current_position': None,
      'current_position_updated': None,
      'duration': None,
      'image_url': None,
      'media_id': None,
      'options': list([
      ]),
      'queue_id': None,
      'song': None,
      'source_id': None,
      'station': None,
      'supported_controls': list([
      ]),
      'type': None,
    }),
    'playback_error': None,
    'player_id': 1,
    'repeat': <RepeatType.OFF: 'off'>,
    'serial': '1234567890',
    'shuffle': False,
    'state': None,
    'supported_version': False,
    'version': '1.493.180',
    'volume': 0,
  })
# ---
# name: test_from_data_version[100.0.0]
  dict({
    'available': True,
    'control': <VolumeControlType.UNKNOWN: 0>,
    'group_id': None,
    'ip_address': '192.168.0.1',
    'is_muted': False,
    'line_out': <LineOutLevelType.VARIABLE: 1>,
    'model': 'HEOS Drive',
    'name': 'Back Patio',
    'network': <NetworkType.WIRED: 'wired'>,
    'now_playing_media': dict({
      'album': None,
      'album_id': None,
      'artist': None,
      'current_position': None,
      'current_position_updated': None,
      'duration': None,
      'image_url': None,
      'media_id': None,
      'options': list([
      ]),
      'queue_id': None,
      'song': None,
      'source_id': None,
      'station': None,
      'supported_controls': list([
      ]),
      'type': None,
    }),
    'playback_error': None,
    'player_id': 1,
    'repeat': <RepeatType.OFF: 'off'>,
    'serial': '1234567890',
    'shuffle': False,
    'state': None,
    'supported_version': True,
    'version': '100.0.0',
    'volume': 0,
  })
# ---
# name: test_from_data_version[3.34.0]
  dict({
    'available': True,
    'control': <VolumeControlType.UNKNOWN: 0>,
    'group_id': None,
    'ip_address': '192.168.0.1',
    'is_muted': False,
    'line_out': <LineOutLevelType.VARIABLE: 1>,
    'model': 'HEOS Drive',
    'name': 'Back Patio',
    'network': <NetworkType.WIRED: 'wired'>,
    'now_playing_media': dict({
      'album': None,
      'album_id': None,
      'artist': None,
      'current_position': None,
      'current_position_updated': None,
      'duration': None,
      'image_url': None,
      'media_id': None,
      'options': list([
      ]),
      'queue_id': None,
      'song': None,
      'source_id': None,
      'station': None,
      'supported_controls': list([
      ]),
      'type': None,
    }),
    'playback_error': None,
    'player_id': 1,
    'repeat': <RepeatType.OFF: 'off'>,
    'serial': '1234567890',
    'shuffle': False,
    'state': None,
    'supported_version': True,
    'version': '3.34.0',
    'volume': 0,
  })
# ---
# name: test_from_data_version[None]
  dict({
    'available': True,
    'control': <VolumeControlType.UNKNOWN: 0>,
    'group_id': None,
    'ip_address': '192.168.0.1',
    'is_muted': False,
    'line_out': <LineOutLevelType.VARIABLE: 1>,
    'model': 'HEOS Drive',
    'name': 'Back Patio',
    'network': <NetworkType.WIRED: 'wired'>,
    'now_playing_media': dict({
      'album': None,
      'album_id': None,
      'artist': None,
      'current_position': None,
      'current_position_updated': None,
      'duration': None,
      'image_url': None,
      'media_id': None,
      'options': list([
      ]),
      'queue_id': None,
      'song': None,
      'source_id': None,
      'station': None,
      'supported_controls': list([
      ]),
      'type': None,
    }),
    'playback_error': None,
    'player_id': 1,
    'repeat': <RepeatType.OFF: 'off'>,
    'serial': '1234567890',
    'shuffle': False,
    'state': None,
    'supported_version': False,
    'version': None,
    'volume': 0,
  })
# ---
# name: test_from_data_version[]
  dict({
    'available': True,
    'control': <VolumeControlType.UNKNOWN: 0>,
    'group_id': None,
    'ip_address': '192.168.0.1',
    'is_muted': False,
    'line_out': <LineOutLevelType.VARIABLE: 1>,
    'model': 'HEOS Drive',
    'name': 'Back Patio',
    'network': <NetworkType.WIRED: 'wired'>,
    'now_playing_media': dict({
      'album': None,
      'album_id': None,
      'artist': None,
      'current_position': None,
      'current_position_updated': None,
      'duration': None,
      'image_url': None,
      'media_id': None,
      'options': list([
      ]),
      'queue_id': None,
      'song': None,
      'source_id': None,
      'station': None,
      'supported_controls': list([
      ]),
      'type': None,
    }),
    'playback_error': None,
    'player_id': 1,
    'repeat': <RepeatType.OFF: 'off'>,
    'serial': '1234567890',
    'shuffle': False,
    'state': None,
    'supported_version': False,
    'version': '',
    'volume': 0,
  })
# ---
# name: test_get_queue
  list([
    dict({
      'album': '22 Break',
      'album_id': '199555605',
      'artist': 'Oh Wonder',
      'image_url': 'http://resources.wimpmusic.com/images/bdfd93c2/0b3a/495e/a557/4493fcbb7ab3/640x640.jpg',
      'media_id': '199555606',
      'queue_id': 1,
      'song': 'Baby',
    }),
    dict({
      'album': '22 Break',
      'album_id': '199555605',
      'artist': 'Oh Wonder',
      'image_url': 'http://resources.wimpmusic.com/images/bdfd93c2/0b3a/495e/a557/4493fcbb7ab3/640x640.jpg',
      'media_id': '199555607',
      'queue_id': 2,
      'song': 'Down',
    }),
    dict({
      'album': '22 Break',
      'album_id': '199555605',
      'artist': 'Oh Wonder',
      'image_url': 'http://resources.wimpmusic.com/images/bdfd93c2/0b3a/495e/a557/4493fcbb7ab3/640x640.jpg',
      'media_id': '199555608',
      'queue_id': 3,
      'song': '22 Break',
    }),
    dict({
      'album': '22 Break',
      'album_id': '199555605',
      'artist': 'Oh Wonder',
      'image_url': 'http://resources.wimpmusic.com/images/bdfd93c2/0b3a/495e/a557/4493fcbb7ab3/640x640.jpg',
      'media_id': '199555609',
      'queue_id': 4,
      'song': 'Free',
    }),
    dict({
      'album': '22 Break',
      'album_id': '199555605',
      'artist': 'Oh Wonder',
      'image_url': 'http://resources.wimpmusic.com/images/bdfd93c2/0b3a/495e/a557/4493fcbb7ab3/640x640.jpg',
      'media_id': '199555610',
      'queue_id': 5,
      'song': "Don't Let The Neighbourhood Hear",
    }),
    dict({
      'album': '22 Break',
      'album_id': '199555605',
      'artist': 'Oh Wonder',
      'image_url': 'http://resources.wimpmusic.com/images/bdfd93c2/0b3a/495e/a557/4493fcbb7ab3/640x640.jpg',
      'media_id': '199555611',
      'queue_id': 6,
      'song': 'Dinner',
    }),
    dict({
      'album': '22 Break',
      'album_id': '199555605',
      'artist': 'Oh Wonder',
      'image_url': 'http://resources.wimpmusic.com/images/bdfd93c2/0b3a/495e/a557/4493fcbb7ab3/640x640.jpg',
      'media_id': '199555612',
      'queue_id': 7,
      'song': 'Rollercoaster Baby',
    }),
    dict({
      'album': '22 Break',
      'album_id': '199555605',
      'artist': 'Oh Wonder',
      'image_url': 'http://resources.wimpmusic.com/images/bdfd93c2/0b3a/495e/a557/4493fcbb7ab3/640x640.jpg',
      'media_id': '199555613',
      'queue_id': 8,
      'song': 'Love Me Now',
    }),
    dict({
      'album': '22 Break',
      'album_id': '199555605',
      'artist': 'Oh Wonder',
      'image_url': 'http://resources.wimpmusic.com/images/bdfd93c2/0b3a/495e/a557/4493fcbb7ab3/640x640.jpg',
      'media_id': '199555614',
      'queue_id': 9,
      'song': 'You > Me',
    }),
    dict({
      'album': '22 Break',
      'album_id': '199555605',
      'artist': 'Oh Wonder',
      'image_url': 'http://resources.wimpmusic.com/images/bdfd93c2/0b3a/495e/a557/4493fcbb7ab3/640x640.jpg',
      'media_id': '199555615',
      'queue_id': 10,
      'song': 'Kicking The Doors Down',
    }),
    dict({
      'album': '22 Break',
      'album_id': '199555605',
      'artist': 'Oh Wonder',
      'image_url': 'http://resources.wimpmusic.com/images/bdfd93c2/0b3a/495e/a557/4493fcbb7ab3/640x640.jpg',
      'media_id': '199555616',
      'queue_id': 11,
      'song': 'Twenty Fourteen',
    }),
  ])
# ---
# name: test_get_queue_with_range
  list([
    dict({
      'album': '22 Break',
      'album_id': '199555605',
      'artist': 'Oh Wonder',
      'image_url': 'http://resources.wimpmusic.com/images/bdfd93c2/0b3a/495e/a557/4493fcbb7ab3/640x640.jpg',
      'media_id': '199555606',
      'queue_id': 1,
      'song': 'Baby',
    }),
    dict({
      'album': '22 Break',
      'album_id': '199555605',
      'artist': 'Oh Wonder',
      'image_url': 'http://resources.wimpmusic.com/images/bdfd93c2/0b3a/495e/a557/4493fcbb7ab3/640x640.jpg',
      'media_id': '199555607',
      'queue_id': 2,
      'song': 'Down',
    }),
    dict({
      'album': '22 Break',
      'album_id': '199555605',
      'artist': 'Oh Wonder',
      'image_url': 'http://resources.wimpmusic.com/images/bdfd93c2/0b3a/495e/a557/4493fcbb7ab3/640x640.jpg',
      'media_id': '199555608',
      'queue_id': 3,
      'song': '22 Break',
    }),
    dict({
      'album': '22 Break',
      'album_id': '199555605',
      'artist': 'Oh Wonder',
      'image_url': 'http://resources.wimpmusic.com/images/bdfd93c2/0b3a/495e/a557/4493fcbb7ab3/640x640.jpg',
      'media_id': '199555609',
      'queue_id': 4,
      'song': 'Free',
    }),
    dict({
      'album': '22 Break',
      'album_id': '199555605',
      'artist': 'Oh Wonder',
      'image_url': 'http://resources.wimpmusic.com/images/bdfd93c2/0b3a/495e/a557/4493fcbb7ab3/640x640.jpg',
      'media_id': '199555610',
      'queue_id': 5,
      'song': "Don't Let The Neighbourhood Hear",
    }),
    dict({
      'album': '22 Break',
      'album_id': '199555605',
      'artist': 'Oh Wonder',
      'image_url': 'http://resources.wimpmusic.com/images/bdfd93c2/0b3a/495e/a557/4493fcbb7ab3/640x640.jpg',
      'media_id': '199555611',
      'queue_id': 6,
      'song': 'Dinner',
    }),
    dict({
      'album': '22 Break',
      'album_id': '199555605',
      'artist': 'Oh Wonder',
      'image_url': 'http://resources.wimpmusic.com/images/bdfd93c2/0b3a/495e/a557/4493fcbb7ab3/640x640.jpg',
      'media_id': '199555612',
      'queue_id': 7,
      'song': 'Rollercoaster Baby',
    }),
    dict({
      'album': '22 Break',
      'album_id': '199555605',
      'artist': 'Oh Wonder',
      'image_url': 'http://resources.wimpmusic.com/images/bdfd93c2/0b3a/495e/a557/4493fcbb7ab3/640x640.jpg',
      'media_id': '199555613',
      'queue_id': 8,
      'song': 'Love Me Now',
    }),
    dict({
      'album': '22 Break',
      'album_id': '199555605',
      'artist': 'Oh Wonder',
      'image_url': 'http://resources.wimpmusic.com/images/bdfd93c2/0b3a/495e/a557/4493fcbb7ab3/640x640.jpg',
      'media_id': '199555614',
      'queue_id': 9,
      'song': 'You > Me',
    }),
    dict({
      'album': '22 Break',
      'album_id': '199555605',
      'artist': 'Oh Wonder',
      'image_url': 'http://resources.wimpmusic.com/images/bdfd93c2/0b3a/495e/a557/4493fcbb7ab3/640x640.jpg',
      'media_id': '199555615',
      'queue_id': 10,
      'song': 'Kicking The Doors Down',
    }),
    dict({
      'album': '22 Break',
      'album_id': '199555605',
      'artist': 'Oh Wonder',
      'image_url': 'http://resources.wimpmusic.com/images/bdfd93c2/0b3a/495e/a557/4493fcbb7ab3/640x640.jpg',
      'media_id': '199555616',
      'queue_id': 11,
      'song': 'Twenty Fourteen',
    }),
  ])
# ---
# name: test_get_quick_selects
  dict({
    1: 'Quick Select 1',
    2: 'Quick Select 2',
    3: 'Quick Select 3',
    4: 'Quick Select 4',
    5: 'Quick Select 5',
    6: 'Quick Select 6',
  })
# ---
# name: test_now_playing_media_unavailable
  dict({
    'album': None,
    'album_id': None,
    'artist': None,
    'current_position': None,
    'current_position_updated': None,
    'duration': None,
    'image_url': None,
    'media_id': None,
    'options': list([
    ]),
    'queue_id': None,
    'song': None,
    'source_id': None,
    'station': None,
    'supported_controls': list([
    ]),
    'type': None,
  })
# ---
# name: test_refresh
  dict({
    'available': True,
    'control': <VolumeControlType.UNKNOWN: 0>,
    'group_id': -263109739,
    'ip_address': '127.0.0.1',
    'is_muted': False,
    'line_out': <LineOutLevelType.VARIABLE: 1>,
    'model': 'HEOS Drive',
    'name': 'Zone 1',
    'network': <NetworkType.WIRED: 'wired'>,
    'now_playing_media': dict({
      'album': 'Album',
      'album_id': '123456',
      'artist': 'Artist',
      'current_position': None,
      'current_position_updated': None,
      'duration': None,
      'image_url': 'http://cont-5.p-cdn.us/images/public/int/6/1/1/9/050087149116_500W_500H.jpg',
      'media_id': '4256592506324148495',
      'options': list([
        dict({
          'context': 'play',
          'id': 11,
          'name': 'Thumbs Up',
        }),
        dict({
          'context': 'play',
          'id': 12,
          'name': 'Thumbs Down',
        }),
        dict({
          'context': 'play',
          'id': 19,
          'name': 'Add to HEOS Favorites',
        }),
      ]),
      'queue_id': 1,
      'song': "Disney (Children's) Radio",
      'source_id': 13,
      'station': "Disney (Children's) Radio",
      'supported_controls': list([
        <ControlType.PLAY: 'play'>,
        <ControlType.PAUSE: 'pause'>,
        <ControlType.STOP: 'stop'>,
        <ControlType.PLAY_NEXT: 'play_next'>,
        <ControlType.PLAY_PREVIOUS: 'play_previous'>,
      ]),
      'type': <MediaType.STATION: 'station'>,
    }),
    'playback_error': None,
    'player_id': -263109739,
    'repeat': <RepeatType.OFF: 'off'>,
    'serial': '123456789',
    'shuffle': False,
    'state': <PlayState.STOP: 'stop'>,
    'supported_version': True,
    'version': '3.34.620',
    'volume': 36,
  })
# ---
# name: test_update_from_data
  dict({
    'available': True,
    'control': <VolumeControlType.UNKNOWN: 0>,
    'group_id': None,
    'ip_address': '192.168.0.2',
    'is_muted': False,
    'line_out': <LineOutLevelType.UNKNOWN: 0>,
    'model': 'HEOS Drives',
    'name': 'Patio',
    'network': <NetworkType.WIFI: 'wifi'>,
    'now_playing_media': dict({
      'album': None,
      'album_id': None,
      'artist': None,
      'current_position': None,
      'current_position_updated': None,
      'duration': None,
      'image_url': None,
      'media_id': None,
      'options': list([
      ]),
      'queue_id': None,
      'song': None,
      'source_id': None,
      'station': None,
      'supported_controls': list([
      ]),
      'type': None,
    }),
    'playback_error': None,
    'player_id': 2,
    'repeat': <RepeatType.OFF: 'off'>,
    'serial': '0987654321',
    'shuffle': False,
    'state': None,
    'supported_version': True,
    'version': '3.34.610',
    'volume': 0,
  })
# ---
